File: frontend/src/engine/capabilities/createChart.ts
/**
 * createChart – Create a chart from data.
 *
 * Office.js notes:
 * - Chart types map to Excel.ChartType enum values.
 * - setData() binds the chart to the source range.
 * - Position is set via left/top/width/height in points.
 * - Charts are added to a sheet's chart collection.
 */

import { CapabilityMeta, CreateChartParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "createChart",
  description: "Create a chart from data range",
  mutates: true,
  affectsFormatting: true,
  requiresApiSet: "ExcelApi 1.1",
};

async function handler(
  context: Excel.RequestContext,
  params: CreateChartParams,
  options: ExecutionOptions
): Promise<StepResult> {
  const { dataRange, chartType, title, sheetName, position, seriesNames } = params;

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would create ${chartType} chart from ${dataRange}`,
    };
  }

  options.onProgress?.(`Creating ${chartType} chart...`);

  // Validate the target sheet if explicitly provided, then resolve it.
  let sheet: Excel.Worksheet;
  if (sheetName) {
    const ws = context.workbook.worksheets.getItemOrNullObject(sheetName);
    ws.load("isNullObject");
    await context.sync();
    if (ws.isNullObject) {
      return {
        stepId: "",
        status: "error",
        message: `Sheet "${sheetName}" not found. Please check the sheet name.`,
      };
    }
    sheet = ws;
  } else {
    sheet = context.workbook.worksheets.getActiveWorksheet();
  }

  const range = resolveRange(context, dataRange);
  const excelChartType = mapChartType(chartType);

  const chart = sheet.charts.add(excelChartType, range, Excel.ChartSeriesBy.auto);

  if (title) {
    chart.title.text = title;
    chart.title.visible = true;
  }

  if (position) {
    chart.left = position.left;
    chart.top = position.top;
    chart.width = position.width;
    chart.height = position.height;
  } else {
    // Default position: below the data
    chart.left = 10;
    chart.top = 300;
    chart.width = 500;
    chart.height = 300;
  }

  // Set series names if provided
  if (seriesNames) {
    chart.series.load("count");
    await context.sync();
    for (let i = 0; i < Math.min(seriesNames.length, chart.series.count); i++) {
      chart.series.getItemAt(i).name = seriesNames[i];
    }
  }

  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Created ${chartType} chart${title ? ` "${title}"` : ""} from ${dataRange}`,
    outputs: { chartName: title ?? `${chartType} chart` },
  };
}

function mapChartType(type: string): Excel.ChartType {
  const map: Record<string, Excel.ChartType> = {
    columnClustered: Excel.ChartType.columnClustered,
    columnStacked:   Excel.ChartType.columnStacked,
    bar:             Excel.ChartType.barClustered,
    line:            Excel.ChartType.line,
    pie:             Excel.ChartType.pie,
    area:            Excel.ChartType.area,
    scatter:         Excel.ChartType.xyscatter,
    combo:           Excel.ChartType.columnClustered, // combo requires special handling
  };
  return map[type] ?? Excel.ChartType.columnClustered;
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/create_chart.py
"""createChart — embed a chart on a sheet from a source range."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


# Map friendly names to xlChartType enum values.
_CHART_TYPE = {
    "column": 51,          # xlColumnClustered
    "columnClustered": 51,
    "bar": 57,             # xlBarClustered
    "line": 4,             # xlLine
    "pie": 5,              # xlPie
    "scatter": -4169,      # xlXYScatter
    "area": 1,             # xlArea
    "doughnut": -4120,     # xlDoughnut
}


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    source = params.get("sourceRange")
    chart_type = (params.get("chartType") or "column").lower()
    title = params.get("title")
    dest_sheet_name = params.get("sheetName")
    position = params.get("position") or {}
    if not source:
        return {"status": "error", "message": "createChart requires 'sourceRange'."}

    book = ctx.workbook_handle
    src = resolve_range(book, source)
    if ctx.dry_run:
        return {"status": "preview", "message": f"Would create {chart_type} chart from {src.address}."}

    try:
        sheet = book.sheets[dest_sheet_name] if dest_sheet_name else src.sheet
        chart = sheet.charts.add(
            left=position.get("left", 100),
            top=position.get("top", 100),
            width=position.get("width", 400),
            height=position.get("height", 300),
        )
        chart.set_source_data(src)
        xl_type_val = _CHART_TYPE.get(chart_type, 51)
        chart.api[1].ChartType = xl_type_val
        if title:
            chart.api[1].HasTitle = True
            chart.api[1].ChartTitle.Text = title
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Create chart failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"Created {chart_type} chart on {sheet.name}.",
        "outputs": {"sheet": sheet.name, "chartType": chart_type},
    }


registry.register("createChart", handler, mutates=True, affects_formatting=True)
